### PR TITLE
update ss3 to 3.30.16, fixes to use NAs in data frames, correct fleet values for OM catch

### PR DESCRIPTION
**Pull request by k-doering-NOAA**
This PR includes several different updates and fixes:
1. Updates the stock synthesis executables to 3.30.16. I ran all tests locally to verify that ss3sim still worked with this new stock synthesis version.
2. Fixes so that NA's can be used in data frames to specify scenarios. This makes it possible to sample a data type for a fleet in 1 scenario, but not another; the NA should be included in the sample size or se/standard dev input for the row specifying the scenario. for example, the following should now work:
```r
library(ss3sim)
df <- setup_scenarios_defaults()
df <- rbind(df, df)
df[2, "sl.Nsamp.2"] <- NA # b/c don't want to sample length comps for fleet 2 in scenario no_lencomps_2
df[, "scenarios"] <- c("ctl", "no_lencomps_2")
df[, "bias_adjust"] <- FALSE
df[, "hess_always"] <- FALSE

run_ss3sim(iterations = 1, simdf = simdf)
```

3. Corrects the fleets that need catch in the operating model (#279 ) . A check was also added to stop the model run if the OM did not include any catch (which suggests something is incorrect about the OM specification)

### CODE REVIEW

File: R/change_catch.R [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#' @param dat_list A data file as read in using `r4ss::SS_readdat()`
+#' @param f_params A list of parameters related to fishing mortality
+#' @return a modified data file as a list in R.
+#' @author Kathryn Doering
+#'
+change_catch <- function(dat_list, f_params) {
+  # Note: would need to modify if did not want catch every year, potentially.
+  dat_list[["catch"]] <- expand.grid(
+    year = c(-999, dat_list$styr:dat_list$endyr),
+    seas = 1, fleet = f_params$fleets, catch = 1, catch_se = 0.01)
```
Comment (kellijohnson-NOAA) [inline R/change_catch.R:12]: @k-doering-NOAA should we think about season being hard-coded here? I am mainly thinking in terms of what you found with Henry's model.

Comment (k-doering-NOAA) [reply R/change_catch.R:12]: good point - should we pull season from the dummy data file instead (and maybe if it is missing, just put the season as 1 with a warning)?

Comment (kellijohnson-NOAA) [reply R/change_catch.R:12]: i think it would be better to have it as part of f_params where the default is one

**Comment by kellijohnson-NOAA** [comment]
Thanks @k-doering-NOAA for these changes. I think that this can be merged into development once we figure out a default way to deal with season that allows users to change season either by fleet or by fleet x year combination. Do we need the latter?

**Comment by k-doering-NOAA** [comment]
Hmm, I'm not sure if the latter is needed - that's a great question. It seems unlikely to be used but possible to specify?

File: R/change_catch.R (hunk 1 above)
Comment (k-doering-NOAA) [reply R/change_catch.R:12]: that's a good idea

**Comment by k-doering-NOAA** [comment]
Thinking more about the seasons issue, it's not immediately clear to me how to solve it. For a multiseason model, it is possible to have catches for each season for each fleet and year. For example, here is an example from the stock synthesis manual of catches in a 2 season model:

```
Catches by year, season for every fleet:
Year Season Fleet Catch Catch SE
-999 1 1 56 0.05
-999 2 1 62 0.05
1975 1 1 876 0.05
1975 2 1 343 0.05
... ... ... ... ...
-999 1 2 55 0.05
-999 2 2 22 0.05
1975 1 2 555 0.05
1975 2 2 873 0.05
... ... ... ... ...
-9999 0 0 0 0
```

I'm not sure what would be a good way to tell ss3sim to generate catches like this through the dataframe approach (for example, here is the default columns related to F/catch produced by `setup_scenario_defaults()`:
```r
cf.years.1      cf.fvals.1 
26:100           rep(0.1052, 75)
```

What if we wanted to specify F's for fleet 1 for both season 1 and 2 (assuming a 2 season model)? Any idea on a way to specify it in this dataframe format?

I suppose this could be an issue for the other specifications (like which data to sample) as well...

**Comment by kellijohnson-NOAA** [comment]
Would you not just supply a season vector for each fleet?
```
cf.years.1       cf.fvals.1                         cf.seas.1
rep(26:100,2)    rep(rep(0.1052, 75),rep(0.2, 75))  rep(1:2, each = 75)
```

**Comment by kellijohnson-NOAA** [comment]
side note, did you know that sample_calcomp uses cpar = 1 when calling sample_comp?

**Comment by kellijohnson-NOAA** [comment]
Regarding sample_calcomp and cpar, I was confused b/c I thought overdispersion was not allowed in the conditional sampling, so cpar should be NA or NULL.

**Comment by k-doering-NOAA** [comment]
@kellijohnson-NOAA not specifying fleets (assuming the same sampling for both fleets) in the dataframe didn't work for me. I think the issue is because a fleets component containing a vector of fleets for the data type is not produced, but ss3sim looks at if this vector of fleets is NULL or has values to decide whether to not sample or sample. I added a test to `test-setup_scenarios` that is currently failing, but I think would pass if setup_scenarios is modified to produced the correct output? 

Another option is modifying the ss3sim_base function to do sampling even if the fleets component for the data type is NULL.

File: R/change_catch.R (hunk 1 above)
Comment (kellijohnson-NOAA) [reply R/change_catch.R:12]: Should we just change line 12 to
```
seas = seq(dat_list[["nseas"]]), fleet = f_params$fleets, catch = 1, catch_se = 0.01)
```

Comment (k-doering-NOAA) [reply R/change_catch.R]: I think a slightly more complicated solution was necessary - see d7c721f . I think we will need to do more thorough testing with seasonal models in the future, but think this PR is ready to be merged.

**Comment by kellijohnson-NOAA** [comment]
Thanks @k-doering-NOAA for all of the work and discussion on this branch.

**Comment by kellijohnson-NOAA** [comment]
@k-doering-NOAA I noticed in the project board you reference something about the scenarios.csv and ensuring it works with two fleets. Can we check that off or is it something we still need to do and should create a new issue/branch/pull request?

**Pull request merged**